### PR TITLE
add support for regex in cache paths

### PR DESCRIPTION
**Pull request by jameslnewell**
Hi,

My app has files named like this:
- `client.[hash].js` 
- `client.[id].[hash].js`

I want to include `client.[hash].js` in my `main` cache and `client.[id].[hash].js` in my `optional` cache. Using a glob makes this ~difficult~ (impossible?) because a glob like `client.*.js` always matches both files.

I want to be able to use a `RegExp` for the cache configuration options in order to get around this problem e.g. `/client.[^.]+.js/`

### CODE REVIEW

**Comment by NekR** [comment]
Hi, @jameslnewell! I'm fine with this change, just need some tests for that to make sure config doesn't break, i.e. glob matching works the same with change and without.

**Comment by jameslnewell** [comment]
🆒  Thanks, will work on that now. Will there be any ramifications of not checking for absolute paths when using a regexp? https://github.com/jameslnewell/offline-plugin/blob/30802324b097ecdbed1cdd3a93f9f1823da045dd/src/index.js#L405

**Comment by NekR** [comment]
I think it should be something like this:

```js
(typeof cacheKey === 'string' || cacheKey instanceof RegExp) &&
!isAbsoluteURL(cacheKey) &&
cacheKey[0] !== '/' &&
cacheKey.indexOf('./') !== 0 &&
(magic = hasMagic(cacheKey))
```
`isAbsoluteURL` and other checks are needed in any case, being it glob pattern or regexp.

**Comment by jameslnewell** [comment]
`cacheKey.indexOf` fails for regex and `cacheKey[0]` returns undefined

**Comment by NekR** [comment]
@jameslnewell yeah, my bad, you got me confused about checking for absolute paths.

There is a check for absolute url only to make sure that glob is valid and to not have false positives in `hasMagic` function (because `http://something` would be weird file system glob).

I think it's mostly fine as you did now. Let's try this:
```js
let magic;

if (typeof cacheKey === 'string') {
  magic =
    !isAbsoluteURL(cacheKey) &&
    cacheKey[0] !== '/' &&
    cacheKey.indexOf('./') !== 0 &&
    hasMagic(cacheKey);
} else if (cacheKey instanceof RegExp) {
  magic = hasMagic(cacheKey);
} else {
  // Ignore non-string and non-RegExp keys
  return;
}

if (magic) {
  // matching with magic...
}
```

File: src/misc/utils.js [hunk 1]
```diff
@@ -49,4 +57,4 @@ export function isAbsoluteURL(url) {
   return /^(?:\w+:)?\/\//.test(url);
 }
 
-export { isAbsolutePath }
\ No newline at end of file
+export { isAbsolutePath }
```
Comment (NekR) [inline src/misc/utils.js]: No line break in the end please here.

File: src/misc/utils.js [hunk 2]
```diff
@@ -6,6 +6,14 @@ const isAbsolutePath = path.isAbsolute;
 // Based on https://github.com/isaacs/node-glob/blob/master/glob.js#L83
 // (glob.hasMagic)
 export function hasMagic(pattern, options) {
+
+  //support RegExp as well as glob
```
Comment (NekR) [inline src/misc/utils.js]: `// Support` and remove empty line between function name and the comment please.

File: tests/legacy/fixtures/paths-regex/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module.exports = __CONFIG__({
+  caches: {
+    main: [/\.js$/]
+  },
+  version: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
```
Comment (NekR) [inline tests/legacy/fixtures/paths-regex/webpack.config.js]: `version: '[hash]'` instead.

**Comment by jameslnewell** [comment]
OK how's that? :shipit:

File: docs/caches.md [hunk 4]
```diff
@@ -8,7 +8,7 @@ ___________________________________
 Tells to the plugin what to cache and how.
 
 * `'all'`: means that everything (all the webpack output assets) and URLs listed in `externals` option will be cached on install.
-* `Object`: Object with 3 possible sections (properties) of type `Array<string>`: `main`, `additional`, `optional`. All sections are optional and by default are empty (no assets added).
+* `Object`: Object with 3 possible sections (properties) of type `Array<string|RegExp>`: `main`, `additional`, `optional`. All sections are optional and by default are empty (no assets added).
```
Comment (NekR) [inline docs/caches.md]: `Array<string | RegExp>`

File: docs/options.md [hunk 5]
```diff
@@ -8,7 +8,7 @@ _Also see list of default options [here](https://github.com/NekR/offline-plugin/
 Allows you to define what to cache and how.
 
 * `'all'`: means that everything (all the webpack output assets) and URLs listed in `externals` option will be cached on install.
-* `Object`: Object with 3 possible `Array<string>` sections (properties): `main`, `additional`, `optional`. All sections are optional and by default are empty (no assets added).
+* `Object`: Object with 3 possible `Array<string|RegExp>` sections (properties): `main`, `additional`, `optional`. All sections are optional and by default are empty (no assets added).
```
Comment (NekR) [inline docs/options.md]: `Array<string | RegExp>`

File: docs/caches.md [hunk 6]
```diff
@@ -43,4 +43,4 @@ AppCache doesn't support conditional or delayed assets loading and by default ig
  AppCache: {
   caches: ['main', 'additional', 'optional']
  }
-```
\ No newline at end of file
+```
```
Comment (NekR) [inline docs/caches.md]: What is this change? Line break in the end?

Comment (jameslnewell) [inline docs/caches.md]: yeah soz, atom's adding them automatically

Comment (NekR) [inline docs/caches.md]: oh, I see. I don't have any special feeling about trailing like breaks, I just don't like unnecessary/unrelated changes because who knows when you will need to merge things and why they could break then.

**Comment by NekR** [comment]
Tests fail.

**Comment by NekR** [comment]
Okay, thank you @jameslnewell! I think I'll merge it tomorrow because I don't have time right for the final review.

**Comment by jameslnewell** [comment]
👍  thanks

**Comment by NekR** [comment]
Okay, merging this, many thanks, @jameslnewell!

_(I'm not sure when will have a version out with this since all this changes go to v5 and there will be some delay before it's out)_

**Comment by jameslnewell** [comment]
no problem
